src/test/groovy/edu/oregonstate/mist/locations/frontend/LocationResourceTest.groovy

```diff
@@ -51,6 +51,14 @@ class LocationResourceTest {
         assert geoJsonRes.entity.type == 'FeatureCollection'
         assert geoJsonRes.entity.hasProperty('features')
 
+        // Test: out of range lat/lon
+        def outOfRange = resource.list(null, null,
+            null, -100, 200, null, null, null, null, null, null, null, null, null)
+        assert outOfRange.status == 400
+        assert outOfRange.entity.userMessage.contains("Bad Request")
+        assert outOfRange.entity.developerMessage.contains("Longitude/latitude out of bounds")
+        assert outOfRange.entity.code == 1400
+
```

Comment:
* Could you also add a test case when `lon` and `lat` are both in this range?
* When I sent this query: `/locations?lat=string&lon=string&distance=1&distanceUnit=mi`, I got this error:

    ```
    {
        "code": 404,
        "message": "HTTP 404 Not Found"
    }
    ```
    Could you validate the types for `lon` and `lat` as well? If they are not numbers, I think we should return a `400` instead of a `404`. What do you think?

Reply:
Sounds good, I'll do that.

Reply:
I discussed this with @ianbrown9475  in person. Currently, we [validate the parameters](https://github.com/osu-mist/locations-frontend-api/blob/develop/src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy#L107-L110) and return a 404 if the parameter is not valid. However, I don't think it is proper cause the 404 should only stand for [the single resource is not found](http://jsonapi.org/format/#fetching-resources-responses-404). I think returning a 400 here make more sense to me. @jaredkosanovic what do you think?

Reply:
@ianbrown9475 let's not doing the validation for now. After you add the test case for valid `lon` and `lat`, could you assign the ticket to @jaredkosanovic and wait for his response?

Reply:
@ianbrown9475 and I ran into another odd issue that when including `lat` and `lon` with the valid values, we got 500 error and the following are the error messages:

```
ERROR [2018-06-20 21:50:06,740] edu.oregonstate.mist.locations.frontend.resources.LocationResource: Exception while getting locations.
! org.elasticsearch.index.query.QueryParsingException: failed to find geo_point field [attributes.geoLocation]
! at org.elasticsearch.index.query.GeoDistanceQueryParser.parse(GeoDistanceQueryParser.java:175)
! at org.elasticsearch.index.query.QueryParseContext.parseInnerQuery(QueryParseContext.java:250)
! at org.elasticsearch.index.query.QueryParseContext.parseInnerFilter(QueryParseContext.java:263)
! at org.elasticsearch.index.query.BoolQueryParser.parse(BoolQueryParser.java:91)
! at org.elasticsearch.index.query.QueryParseContext.parseInnerQuery(QueryParseContext.java:250)
! at org.elasticsearch.index.query.IndexQueryParserService.innerParse(IndexQueryParserService.java:324)
! at org.elasticsearch.index.query.IndexQueryParserService.parse(IndexQueryParserService.java:224)
! at org.elasticsearch.index.query.IndexQueryParserService.parse(IndexQueryParserService.java:219)
! at org.elasticsearch.search.query.QueryParseElement.parse(QueryParseElement.java:33)
! at org.elasticsearch.search.SearchService.parseSource(SearchService.java:856)
! at org.elasticsearch.search.SearchService.createContext(SearchService.java:667)
! at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:633)
! at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:377)
! at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:368)
! at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:365)
! at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
! at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:77)
! at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:378)
! at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
! ... 3 common frames omitted
! Causing: org.elasticsearch.action.search.SearchPhaseExecutionException: all shards failed
! at org.elasticsearch.action.search.AbstractSearchAsyncAction.onFirstPhaseResult(AbstractSearchAsyncAction.java:206)
! at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:152)
! at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:46)
! at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:874)
! at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:852)
! at org.elasticsearch.transport.TransportService$4.onFailure(TransportService.java:389)
! at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
! at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
! at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
! at java.lang.Thread.run(Thread.java:748)
127.0.0.1 - - [20/Jun/2018:21:50:06 +0000] "GET /api/v1/locations?lat=44.564917&lon=-123.278976&distance=200&distanceUnit=yd&type=parking HTTP/1.1" 500 285 "-" "PostmanRuntime/7.1.1" 367
```

It only happens when running API locally but I've never seen this error before. Sending the same query to prod and dev works fine. I've checked the locations data, ES version, and the API branch. All of them are exactly same between local API and dev/prod API. Not sure what's happening. @jaredkosanovic have you run into this issue before?

Reply:
@tsoliangwu0130 Regarding the exception you ran into, I wasn't able to replicate that locally. One thing that stuck out was this line
`! org.elasticsearch.index.query.QueryParsingException: failed to find geo_point field [attributes.geoLocation]`

perhaps the json in your local es index is corrupted?

Regarding returning a 400 instead of a 404, I agree. 404 doesn't make much sense here. However, changing that status code could be a breaking change for consuming applications, who might be expecting that 404 response code for their application. It makes me wonder if we should start collecting some of these changes and thinking about releasing a v2 of the locations api. It was one of our first APIs, and I think we could be able to come up with some improvements and might require it to be versioned. Either way, that's a separate ticket/pr :)

Reply:
@jaredkosanovic Which json did you use? I got the issue using the json in [api-dev](http://act-jenkins.ucsadm.oregonstate.edu:8080/view/APIs/job/apis%20Locations%20API%20Backend%20ES%20Update%20%28api-dev%29) and [api-prod](http://act-jenkins.ucsadm.oregonstate.edu:8080/view/APIs/job/apis%20Locations%20API%20Backend%20ES%20Update%20%28api-prod%29) from the Jenkins links in the frontend readme. I think @tsoliangwu0130 also used these but may have tried some others too, not sure.

Reply:
I'm fine with not to change it for now. But yeah, we should definitely update our APIs + skeleton since there are too many things need to be changed, for example, update swagger to OpenAPI 3.0, make sure our skeleton really follow JSON API, etc. Do you have any preferred time to do this? Could you create a ticket for that?